### ISSUE TITLE
Performance of streaming API

### ISSUE THREAD

**Issue by alexkamp** [msgpack/msgpack-c#746]
I have an application which receives msgpacked data in chunks of approx. 20 bytes. I can not guarantee that message boundaries are preserved, meaning that e.g. a msgpack array which has a length of 35 bytes may be split in two messages. In order to handle this, I use code like this: 

```
static size_t refill(msgpack_unpacker *upk) {
    msgpack_unpacker_reserve_buffer(upk, BUFFERSIZE);

   // receive data ...
   
    uint8_t *buf = (uint8_t *) msgpack_unpacker_buffer(upk);
    memcpy(buf, reveived_data, received_length);

    msgpack_unpacker_buffer_consumed(upk, received_length);
    return received_length;
}
do {
  ret = msgpack_unpacker_next(&upk, &result);
  if(MSGPACK_UNPACK_SUCCESS != ret && MSGPACK_UNPACK_EXTRA_BYTES != ret) {
    refill(&upk);
  }
} while(ret != MSGPACK_UNPACK_SUCCESS && ret != MSGPACK_UNPACK_EXTRA_BYTES);
```

I attached a reproducible example, which simulates the receiving of data by copying from a buffer.
[msgpack-perf.tar.gz](https://github.com/msgpack/msgpack-c/files/2607962/msgpack-perf.tar.gz)

Applying [google-pprof](https://github.com/google/pprof) to the attached example, I can see that 90% of the runtime are spend in msgpack_unpacker_next, with 42% in msgpack_unpacker_release_zone. Apparently, msgpack_unpacker_next frees and re-allocates the msgpack_zone on every invocation. 

Is there any possibility to speed this up? It seems unnecessarily slow. 

The attached tar archive contains my example program, a script to compile and run it and the profiling data I collected, including the pdf representation of the data.

**Comment by redboltz** [comment]
Hmm... 
Each `msgpack_unpacked` has `msgpack_zone`. If you use `ARRAY`, all data is unpacked on one `msgpack_zone`.

I mean something like this:

```cpp
    msgpack_packer* pk = msgpack_packer_new(buffer, msgpack_sbuffer_write);
    msgpack_pack_array(pk, TESTSIZE);
    for(uint64_t i = 0; i < TESTSIZE;i++) {
        msgpack_pack_uint64(pk, i);
    }
```

After this modification, `msgpack_unpacker_next()` returns `MSGPACK_UNPACK_SUCCESS` only if the all elements of the `ARRAY` is processed. That doesn't mean msgpack-c do nothing until all buffer is filled. msgpack-c unpacks buffer partially and store the context, and in the next `msgpack_unpacker_next()` call, restore the context and continues unpacking process.That means unpacking load is scatterd.

However, you might want to do application process partially unpacked data (each uint64 data).
For that purpose, msgpack-c provides visitor API. https://github.com/msgpack/msgpack-c/wiki/v2_0_cpp_visitor
It's only for C++.

Please serach `You can use visitors with stream interface like an unpacker.` in the wiki page. You can find the way to combine stream input (unpredictable length) with visitor API.

If you use the visitor API, you can do anything what you want with the stream directly. There is no zone allocation. That means you don't need to use `ARRAY` any more. It might be a good customization point for you.

Unfortunately, I have no plan to support visitor interface for C.

**Comment by alexkamp** [comment]
Could you elaborate on why the unpacker_next API re-allocates the zone every time? It looks like it could just reuse the existing object?

**Comment by redboltz** [comment]
> Could you elaborate on why the unpacker_next API re-allocates the zone every time? It looks like it could just reuse the existing object?

Consider the following scenario:

1. `msgpack_unapcker_next()` creates huge object. It allocate huge zone.
2. Call `msgpack_unpacker_next()` again, and if reused the object `msgpack_unpacked`, memory usage is still high. It's not good.

Reallocating zone is for shrink memory usage.

I'm not original version author, so I'm not 100% sure the design policy this part. But maybe it is the reason.

**Comment by alexkamp** [comment]
Yes, that is most like the reason for the current implementation. 
However, the zone is only used for arrays and maps. My examples uses uint64 only, so the zone will never be used. If I disable zone recreation, my example runs in 5 seconds, instead of 12 seconds. Obviously this version will fail on maps and arrays, as the zone is not even allocated. 

Do you think it would be possible to check whether the zone has been used, and recreate it only if it was?

**Comment by redboltz** [comment]
> However, the zone is only used for arrays and maps. 

Good point. I will consider how can I reuse the zone if the object is neither ARRAY nor MAP.

NOTE: 
STR, BIN, and EXT is shallow copy (pointer copy) in C version.
C++ version can choose shallow copy or deep copy (to zone). And the default behavior is deep copy.
It's inconsistent.

Anyway, I will try to update C version.

**Comment by redboltz** [comment]
I noticed that zone has the reference counter decrement function pointer in its finalizer array if the object type is STR, BIN, or EXT. 

(Reference counter is for unpacker buffer.)

See the second diagram BIN https://github.com/msgpack/msgpack-c/wiki/v2_0_cpp_object
The diagram describes C++ internal structure, but BIN choice is the same as C.

So reusable case is except ARRAY, MAP, STR, BIN, and EXT.

**Comment by redboltz** [comment]
@alexkamp , I implemented #747 . Could you try this on your environment?

**Comment by redboltz** [comment]
Thank you for checking. Merged.
